drv/mb85rsxx-fram/src/lib.rs

```diff
@@ -0,0 +1,489 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! A driver for the Fujitsu MB85RS series of SPI ferroelectric RAM (FRAM)
+//! chips.
+//!
+//! See
+//! <https://www.mouser.com/datasheet/2/1113/MB85RS64T_DS501_00051_2v0_E-2329177.pdf>
+
+#![no_std]
+
+use bitflags::bitflags;
+use drv_spi_api::{CsState, SpiDevice, SpiError, SpiServer};
+use ringbuf::ringbuf_entry;
+
+pub type Mb85rs64v<S> = Fram<S, { product_id::MB85RS64V }>;
+pub type Mb85rs64t<S> = Fram<S, { product_id::MB85RS64T }>;
+pub type Mb85rs256ty<S> = Fram<S, { product_id::MB85RS256TY }>;
+pub type Mb85rs1mt<S> = Fram<S, { product_id::MB85RS1MT }>;
+pub type Mb85rs2mta<S> = Fram<S, { product_id::MB85RS2MTA }>;
+pub type Mb85rs4mt<S> = Fram<S, { product_id::MB85RS4MT }>;
+
+/// A generic Fujitsu FRAM chip of arbitrary size.
+///
+/// By default, the write enable latch on the FRAM is not set, so this type
+/// cannot be written to. To write to a FRAM chip, first call
+/// [`Fram::write_enable`], which returns a [`WritableFram`].
+#[must_use = "a Fram does nothing if constructed but not read from or written to"]
+pub struct Fram<S: SpiServer, const ID: u16> {
+    spi: SpiDevice<S>,
+}
+
+/// A generic Fujitsu FRAM chip with its write enable latch set.
+///
+/// This type is returned by [`Fram::write_enable`], and will unset the write
+/// latch when it's dropped. This way, the FRAM remains in the write-protected
+/// state when you're not actively trying to write to it.
+///
+/// To write to the FRAM chip, use [`WritableFram::write`]. This type also
+/// exposes a [`WritableFram::read`] method, so the FRAM chip may also be read
+/// from while the write enable latch is set.
+#[must_use = "a WritableFram does nothing if constructed but not read from \
+    or written to"]
+pub struct WritableFram<'fram, S: SpiServer, const ID: u16>(
+    &'fram Fram<S, { ID }>,
```

Comment:
Yeah...my initial design for this took the `Fram` by value, but unfortunately, since `WritableFram` has a drop impl and `SpiDevice` is not `Copy` or `Clone`, there isn't a nice  way to get the `Fram` back _out_ of the `WritableFram` besides using `ManuallyDrop`, which I didn't really want to do...but maybe I should have. I'll change it, thanks.